### ISSUE TITLE
Allow torch.nn.Linear to have 0 in_features

### ISSUE THREAD

**Issue by carlosgmartin** [pytorch/pytorch#48152]
## 🚀 Feature

Allow [`torch.nn.Linear`](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) to have 0 `in_features`.

Currently, it throws a `ZeroDivisionError` due to the way weights are initialized. See the [`reset_parameters`](https://pytorch.org/docs/stable/_modules/torch/nn/modules/linear.html#Linear) method.

## Motivation

I'm working in a setting where I automatically construct neural networks according to the input features of a given problem. Sometimes, such a problem has no input features at all. For example, in a pure [normal-form game](https://en.wikipedia.org/wiki/Normal-form_game) or [multi-armed bandit problem](https://en.wikipedia.org/wiki/Multi-armed_bandit), there are no observations, and the network simply has to learn a fixed vector of logits. This would correspond to the bias vector of the `torch.nn.Linear`, if it wasn't throwing an error.

Creating a special case for such situations results in cumbersome code.

## Pitch

PyTorch already supports arithmetic with dimensions of size 0:

```python3
import torch
n_features = 0
n_actions = 5
weights = torch.rand(n_actions, n_features)
bias = torch.rand(n_actions)
features = torch.rand(n_features)
logits = weights @ features + bias
```

Fixing this issue would simply require adjusting `reset_parameters`.

cc @albanD @mruberry @jbschlosser

**Comment by mruberry** [comment]
Thanks for this suggestion, @carlosgmartin. We would take a PR fixing this as long as there were no BC or performance implications.

**Comment by carlosgmartin** [comment]
@mruberry Thank you. Would it be possible to fix the same problem in the RNN modules, e.g. in [`RNNBase.reset_parameters`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/rnn.py#L194) and [`RNNCellBase.reset_parameters`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/rnn.py#L885), which have the line
```
stdv = 1.0 / math.sqrt(self.hidden_size)
```
Should I open a new issue for that?

**Comment by mruberry** [comment]
@carlosgmartin new issue is a great idea ;)
